File: java/rotational-cipher/src/main/java/RotationalCipher.java
class RotationalCipher {

    private static final String ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private int shiftKey;

    RotationalCipher(int shiftKey) {
        this.shiftKey = shiftKey;
    }

    String rotate(String data) {
        StringBuilder sb = new StringBuilder();
        data.chars().forEach(ch -> sb.append(rotateChar((char) ch)));
        return sb.toString();
    }

    private char rotateChar(char ch) {
        Character uc = Character.toUpperCase(ch);
        int i = ALPHABET.indexOf(uc);
        if (i == -1) return ch;

        int shifted = (i + shiftKey) % ALPHABET.length();
        char coded = ALPHABET.charAt(shifted);
        return (ch == uc ? coded : Character.toLowerCase(coded));
    }
}

File: typescript/rotational-cipher/rotational-cipher.ts
const ALPHABET = 'abcdefghijklmnopqrstuvwxyz'

export function rotate(input: string, rotation: number = 13): string {
  return [...input].map((c) => rotateChar(c, rotation)).join('')
}

function rotateChar(c: string, n: number): string {
  const lc = c.toLowerCase()
  const i = ALPHABET.indexOf(lc)
  if (i === -1) { return c }
  const d = ALPHABET[ (i + n) % ALPHABET.length ]
  return c === lc ? d : d.toUpperCase()
}
